Lind-Project/safeposix-rust
# Add tests and comments for select_syscall

## Description

This PR added more tests for the select syscall, and marked some TODO and BUG in existing select syscalls. Tests being added including:
1. test for select monitoring on five different file descriptors:
- Regular file
- AF_INET server socket waiting for two clients
- AF_INET server socket's connection file descriptor with clients
- AF_UNIX server socket's connection file descriptor with a client
- Pipe
2. test for select working with pipe writefd when pipe is full
3. test for select working with AF_UNIX socket pipe writefd when pipe is full
- **This test currently will fail since select did not check for this situation. BUG is marked in select_writefds and this tests is currently commented out**
4. test for timeout field of select, two sub-tests were conducted here:
- select when timeout could expire
- select when all arguments were None except for timeout
5. test for error case of select
- **test for sending signal to select is marked as TODO since sending signals using kill_syscall is currently not supported in raw safeposix environment**

Marked BUG including:
1. under select_writefds, AF_UNIX socket condition, socket's send pipe is not checked for writing, causing one of the test fail
2. under select_writefds, sockets are always considered to be writable, which is not true since socket could block in cases like the kernel send buffer is full.

Marked TODO including:
1. select for exceptfds: we currently don't implement exceptfds and possibly could if necessary
2. select test with signals are marked as to do since sending signals using kill_syscall is currently not supported in raw safeposix environment

Besides, more comments were added to select_syscalls and its sub-functions.

Other noteworthy things:
1. Some tests used non-necessary barriers to make sure some codes are running in order, this is because there are some unfixed issues (pipe reference counter, etc) that could cause the tests fail. Using extra barrier is a workaround to skip these unexpected unfixed issues and make the test work normally.

### Type of change




## How Has This Been Tested?

This PR is addition of new tests and currently all tests were passed.

## Checklist:



- [x] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)
